File: packages/core-ts/src/storage/session-lock.ts
/**
 * File-based session lock to prevent duplicate relay sessions.
 *
 * When multiple MCP server processes start simultaneously (common in Claude Code),
 * this lock prevents each from creating a separate relay session (rate limit:
 * max 10 sessions/IP/10min).
 *
 * Lock file location: <config_dir>/mcp/relay-session-<server>.lock
 */

import { existsSync } from 'node:fs'
import { mkdir, readFile, rename, unlink, writeFile } from 'node:fs/promises'
import { dirname, join } from 'node:path'
import envPaths from 'env-paths'

const paths = envPaths('mcp', { suffix: '' })
const DEFAULT_LOCK_DIR = paths.config

const DEFAULT_MAX_AGE_MS = 600_000

// Allow overriding lock directory for testing
let lockDirOverride: string | null = null

export function setLockDir(path: string | null): void {
  lockDirOverride = path
}

function getLockDir(): string {
  return lockDirOverride ?? DEFAULT_LOCK_DIR
}

function lockPath(serverName: string): string {
  return join(getLockDir(), `relay-session-${serverName}.lock`)
}

export interface SessionInfo {
  sessionId: string
  relayUrl: string
  createdAt: number // Date.now()
}

interface SessionInfoJson {
  session_id: string
  relay_url: string
  created_at: number
}

export async function acquireSessionLock(
  serverName: string,
  maxAgeMs = DEFAULT_MAX_AGE_MS
): Promise<SessionInfo | null> {
  const path = lockPath(serverName)
  try {
    if (!existsSync(path)) {
      return null
    }

    const raw = await readFile(path, 'utf-8')
    const data = JSON.parse(raw) as SessionInfoJson

    // Validate required fields
    if (
      typeof data.session_id !== 'string' ||
      typeof data.relay_url !== 'string' ||
      typeof data.created_at !== 'number'
    ) {
      await releaseSessionLock(serverName)
      return null
    }

    const info: SessionInfo = {
      sessionId: data.session_id,
      relayUrl: data.relay_url,
      createdAt: data.created_at
    }

    const age = Date.now() - info.createdAt
    if (age > maxAgeMs) {
      // Expired lock: clean up and return null
      await releaseSessionLock(serverName)
      return null
    }

    return info
  } catch {
    // Corrupt or unreadable lock file: clean up and return null
    try {
      await releaseSessionLock(serverName)
    } catch {
      /* ignore cleanup errors */
    }
    return null
  }
}

export async function writeSessionLock(serverName: string, info: SessionInfo): Promise<void> {
  const path = lockPath(serverName)
  const dir = dirname(path)
  if (!existsSync(dir)) {
    await mkdir(dir, { recursive: true })
  }

  const data: SessionInfoJson = {
    session_id: info.sessionId,
    relay_url: info.relayUrl,
    created_at: info.createdAt
  }

  // Write atomically via temp file + rename
  const tmpPath = `${path}.tmp`
  await writeFile(tmpPath, JSON.stringify(data), 'utf-8')
  await rename(tmpPath, path)
}

export async function releaseSessionLock(serverName: string): Promise<void> {
  const path = lockPath(serverName)
  try {
    if (existsSync(path)) {
      await unlink(path)
    }
  } catch {
    /* ignore errors during cleanup */
  }
}

File: packages/core-py/src/mcp_relay_core/storage/session_lock.py
"""File-based session lock to prevent duplicate relay sessions.

When multiple MCP server processes start simultaneously (common in Claude Code),
this lock prevents each from creating a separate relay session (rate limit:
max 10 sessions/IP/10min).

Lock file location: <config_dir>/mcp/relay-session-<server>.lock
"""

import json
import logging
import time
from dataclasses import dataclass
from pathlib import Path

from platformdirs import user_config_dir

logger = logging.getLogger(__name__)

_CONFIG_DIR = Path(user_config_dir("mcp", appauthor=False))
_DEFAULT_MAX_AGE_S = 600.0

# Allow overriding lock directory for testing
_lock_dir_override: str | None = None


def set_lock_dir(path: str | None) -> None:
    """Override lock directory (for testing). Pass None to reset."""
    global _lock_dir_override
    _lock_dir_override = path


def _get_lock_dir() -> Path:
    if _lock_dir_override is not None:
        return Path(_lock_dir_override)
    return _CONFIG_DIR


def _lock_path(server_name: str) -> Path:
    return _get_lock_dir() / f"relay-session-{server_name}.lock"


@dataclass
class SessionInfo:
    """Information about an active relay session."""

    session_id: str
    relay_url: str
    created_at: float


async def _read_session_lock(server_name: str) -> SessionInfo | None:
    """Read and parse the session lock file.

    Handles corruption by cleaning up the file and returning None.
    """
    path = _lock_path(server_name)
    try:
        if not path.exists():
            return None

        data = json.loads(path.read_text(encoding="utf-8"))
        return SessionInfo(
            session_id=data["session_id"],
            relay_url=data["relay_url"],
            created_at=data["created_at"],
        )
    except (json.JSONDecodeError, KeyError, OSError) as err:
        logger.debug("Invalid session lock for %s: %s", server_name, err)
        # Corrupt lock file: clean up and return None
        try:
            await release_session_lock(server_name)
        except OSError:
            pass
        return None


async def acquire_session_lock(
    server_name: str,
    max_age_s: float = _DEFAULT_MAX_AGE_S,
) -> SessionInfo | None:
    """Try to acquire or reuse an existing session lock.

    Returns existing SessionInfo if lock exists and is fresh (< max_age_s old).
    Returns None if no valid lock exists (caller should create new session).
    Uses atomic file operations for safety.

    Args:
        server_name: Server identifier.
        max_age_s: Maximum age of a valid lock in seconds.

    Returns:
        SessionInfo if a valid lock exists, None otherwise.
    """
    info = await _read_session_lock(server_name)
    if not info:
        return None

    age = time.time() - info.created_at
    if age > max_age_s:
        logger.debug(
            "Session lock for %s expired (age=%.1fs > max=%.1fs)",
            server_name,
            age,
            max_age_s,
        )
        # Expired lock: clean up and return None
        await release_session_lock(server_name)
        return None

    logger.debug(
        "Reusing session lock for %s (age=%.1fs)",
        server_name,
        age,
    )
    return info


async def write_session_lock(server_name: str, info: SessionInfo) -> None:
    """Write session info to lock file after creating a new relay session.

    Args:
        server_name: Server identifier.
        info: Session information to persist.
    """
    path = _lock_path(server_name)
    path.parent.mkdir(parents=True, exist_ok=True)

    data = {
        "session_id": info.session_id,
        "relay_url": info.relay_url,
        "created_at": info.created_at,
    }

    # Write atomically via temp file + rename
    tmp_path = path.with_suffix(".tmp")
    tmp_path.write_text(json.dumps(data), encoding="utf-8")
    tmp_path.replace(path)

    logger.debug("Wrote session lock for %s", server_name)


async def release_session_lock(server_name: str) -> None:
    """Remove the lock file (call when session completes or is abandoned).

    Args:
        server_name: Server identifier.
    """
    path = _lock_path(server_name)
    try:
        path.unlink(missing_ok=True)
        logger.debug("Released session lock for %s", server_name)
    except OSError as err:
        logger.debug("Failed to release session lock for %s: %s", server_name, err)
